src/ray/raylet/node_manager.cc

```diff
@@ -385,11 +381,27 @@ void NodeManager::HeartbeatAdded(gcs::AsyncGcsClient *client, const ClientID &cl
   remote_resources.SetAvailableResources(std::move(remote_available));
   // Extract the load information and save it locally.
   remote_resources.SetLoadResources(std::move(remote_load));
+}
 
-  auto decision = scheduling_policy_.SpillOver(remote_resources);
+void NodeManager::HeartbeatBatchAdded(const HeartbeatBatchTableDataT &heartbeat_batch) {
+  const ClientID &local_client_id = gcs_client_->client_table().GetLocalClientId();
+  std::vector<ClientID> client_ids(heartbeat_batch.batch.size());
+  // Update load information provided by each heartbeat.
+  for (const auto &heartbeat_data : heartbeat_batch.batch) {
+    const ClientID &client_id = ClientID::from_binary(heartbeat_data->client_id);
+    if (client_id == local_client_id) {
+      // Skip heartbeats from self.
+      continue;
+    }
+    HeartbeatAdded(client_id, *heartbeat_data);
+    client_ids.push_back(client_id);
+  }
   // Extract decision for this local scheduler.
+  auto decision = scheduling_policy_.SpillOver(cluster_resource_map_, client_ids);
```

Comment:
OK, this is where the problem is introduced. You don't want to spillover to the entire cluster. You ONLY want to spillover to the heartbeating raylets. Similarly to how the cluster map was constructed in the original HeartbeatAdded code, you need to construct a cluster map for a heartbeating batch of nodes and pass _that_ to SpillOver, not the whole cluster resource map.

Reply:
I passed in the cluster map just so I didn't have to deal with copying KV pairs to a new map. In the function I only use the keys corresponding to `client_ids` which does what you're describing I think.